lightning/src/routing/onion_message.rs

```diff
@@ -0,0 +1,328 @@
+		match visited.entry(node_id) {
+			hash_map::Entry::Occupied(_) => continue,
+			hash_map::Entry::Vacant(e) => e.insert(parent_node_id),
+		};
+		if node_id == dest_node_id {
+			return Ok(reverse_path(visited, our_node_id, dest_node_id, logger)?)
+		}
+		if let Some(node_info) = network_nodes.get(&node_id) {
+			// Only consider the network graph if first_hops does not override it.
+			if valid_first_hops.contains(&node_id) || node_id == our_node_id {
+			} else if let Some(node_ann) = &node_info.announcement_info {
+				if !node_ann.features.supports_onion_messages() || node_ann.features.requires_unknown_bits()
+				{ continue; }
+			}
+			for scid in &node_info.channels {
+				if let Some(chan_info) = network_channels.get(&scid) {
+					if let Some((directed_channel, successor)) = chan_info.as_directed_from(&node_id) {
+						if let Some(direction) = directed_channel.direction() {
+							if direction.enabled {
+								// We may push a given successor multiple times, but the heap should sort its best
+								// entry to the top. We do this because there is no way to adjust the priority of an
+								// existing entry in `BinaryHeap`.
+								frontier.push(PathBuildingHop {
+									cost: cost + 1,
+									node_id: *successor,
+									parent_node_id: node_id,
+								});
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	Err(Error::PathNotFound)
+}
+
+#[derive(Debug, PartialEq)]
+/// Errored running [`find_path`].
+pub enum Error {
+	/// We failed to find a path to the destination.
+	PathNotFound,
+	/// We failed to convert this node id into a [`PublicKey`].
+	InvalidNodeId(secp256k1::Error),
+}
+
+#[derive(Eq, PartialEq)]
+struct PathBuildingHop {
+	cost: u64,
+	node_id: NodeId,
+	parent_node_id: NodeId,
+}
+
+impl PartialOrd for PathBuildingHop {
+	fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+		// We need a min-heap, whereas `BinaryHeap`s are a max-heap, so compare the costs in reverse.
+		// TODO: randomize which node is selected if they are equal-cost. Currently there is an
+		// incentive to grind lower node ids to attract more message forwards.
+		other.cost.partial_cmp(&self.cost)
+	}
+}
+
+impl Ord for PathBuildingHop {
+	fn cmp(&self, other: &Self) -> cmp::Ordering {
+		self.partial_cmp(other).unwrap()
+	}
+}
+
+fn reverse_path<L: Deref>(
+	parents: HashMap<NodeId, NodeId>, our_node_id: NodeId, destination: NodeId, logger: L
+)-> Result<Vec<PublicKey>, Error> where L::Target: Logger
+{
+	let mut path = Vec::new();
+	let mut curr = destination;
+	loop {
+		match PublicKey::from_slice(curr.as_slice()) {
+			Ok(pk) => path.push(pk),
+			Err(e) => return Err(Error::InvalidNodeId(e))
+		}
+		match parents.get(&curr) {
+			None => return Err(Error::PathNotFound),
+			Some(parent) => {
+				if *parent == our_node_id { break; }
+				curr = *parent;
+			}
+		}
+	}
+
+	path.reverse();
+	log_info!(logger, "Got route to {:?}: {:?}", destination, path);
```

Comment:
As `reverse_path` is kind of a helper function, maybe this log could be placed in `find_route` itself?